datafusion/common/src/scalar.rs

```diff
@@ -359,29 +359,77 @@ impl PartialOrd for ScalarValue {
             (FixedSizeBinary(_, _), _) => None,
             (LargeBinary(v1), LargeBinary(v2)) => v1.partial_cmp(v2),
             (LargeBinary(_), _) => None,
-            (List(arr1), List(arr2))
-            | (FixedSizeList(arr1), FixedSizeList(arr2))
-            | (LargeList(arr1), LargeList(arr2)) => {
-                // ScalarValue::List / ScalarValue::FixedSizeList / ScalarValue::LargeList are ensure to have length 1
+            // ScalarValue::List / ScalarValue::FixedSizeList / ScalarValue::LargeList are ensure to have length 1
+            (List(arr1), List(arr2)) => {
                 assert_eq!(arr1.len(), 1);
                 assert_eq!(arr2.len(), 1);
 
                 if arr1.data_type() != arr2.data_type() {
                     return None;
                 }
 
-                fn first_array_for_list(arr: &ArrayRef) -> ArrayRef {
-                    if let Some(arr) = arr.as_list_opt::<i32>() {
-                        arr.value(0)
-                    } else if let Some(arr) = arr.as_list_opt::<i64>() {
-                        arr.value(0)
-                    } else if let Some(arr) = arr.as_fixed_size_list_opt() {
-                        arr.value(0)
-                    } else {
-                        unreachable!("Since only List / LargeList / FixedSizeList are supported, this should never happen")
+                fn first_array_for_list(arr: &Arc<ListArray>) -> ArrayRef {
```

Comment:
It would be really nice here to avoid the redundant copy/paste code, perhaps by making a function with signature like this

```
fn cmp_list_array(arr1: &dyn Array, arr2: &dyn Array) -> Option<Ordering> {
..
}
```

Which then you can call with `Arc<ListArray>` and `Arc<LargeListArray>` types

Reply:
This one is a little tougher, because if the args to the function are "generic" (i.e. ArrayRef instead of a specific `Arc<ListArray>` type), the `arr1.value(0)` method is not available. The `first_array_for_list` function needs to be able to convert the `ArrayRef` into the specific type to then get the value out. So I'm struggling a bit coming up with a way here.

Reply:
Here is a proposal of how to do it: https://github.com/rspears74/arrow-datafusion/pull/1 (that targets this PR)